src/utils/object.js

```diff
@@ -44,3 +46,65 @@ export function map(opt_initial) {
 export function hasOwn(obj, key) {
   return hasOwn_.call(obj, key);
 }
+
+/**
+ * @param {!Object} target
+ * @param {!Object} source
+ * @param {number} maxDepth The maximum depth for deep merge, beyond which
+ *    Object.assign will be used.
+ * @return {!Object}
+ */
+function deepMerge_(target, source, maxDepth) {
+  // Keep track of seen objects to prevent infinite loops on objects with
+  // recursive references.
+  const seen = [];
+  // Traversal must be breadth-first so any object encountered for the first
+  // time does not have a reference  at a shallower depth. Otherwise, a
+  // recursive reference found at depth == maxDepth could cause an unexpected
+  // change at a shallower depth if the same object exists at a shallower depth.
+  const queue = [{target, source, currentDepth: 0}];
+  while (queue.length > 0) {
+    const {target, source, currentDepth} = queue.shift();
+    if (seen.includes(target) || seen.includes(source)) {
+      // No recursive merge
+      continue;
+    }
+    seen.push(target, source);
```

Comment:
Do you need to keep track of seen objects in `target`? Circular refs must not exist in `source` but is there an issue with allowing circular refs in the bindable scope in general?